neps/nep-0455.md

```diff
@@ -0,0 +1,188 @@
+The chunk should still be included in a later block to avoid stalling the affected shard.
+
+Pros:
+- Fast and automatic adaptation to the blockchain workload
+
+Cons:
+- Needs a voting mechanism to decide that processing of the chunk is slow (as opposed to the chunk producer being slow) and the chunk should be skipped within a block
+- Needs economic design for the rewards for skipped chunks
+- No prior examples of this approach in other blockchains
+
+## Specification
+
+- **Chunk Compute Usage** -- total compute time spent on processing the chunk
+
+- **Chunk Compute Limit** -- upper-bound for compute time spent on processing the chunk
+
+Today, gas has two somewhat orthogonal roles:
+
+1. Gas is money. It is used to avoid spam by charging users
+2. Gas is CPU time. It defines how many transactions fit in a chunk so that validators can apply it within a second
+
+The idea is to decouple these two by introducing parameter compute compute costs.
+Each gas parameter still has a gas cost that determines what users have to pay.
+But when filling a chunk with transactions, parameter compute cost is used to estimate CPU time.
+
+Ideally, all compute costs should match corresponding gas costs.
+But when we discover undercharging issues, we can set a higher compute cost (this would require a protocol upgrade).
+The stability concern is then resolved when the compute cost becomes active.
+
+The ratio between compute cost and gas cost can be thought of as an undercharging factor.
+If a gas cost is 2 times too low to guarantee stability, compute cost will be twice the gas cost.
+A chunk will be full 2 times faster when gas for this parameter is burned.
+This deterministically throttles the throughput to match what validators can actually handle.
+
+Compute costs influence the gas price adjustment logic described in https://nomicon.io/Economics/Economic#transaction-fees.
+Specifically, we're now using compute usage instead of gas usage in the formula to make sure that the gas price increases if chunk processing time is close to the limit.
+
+Compute costs are static for each protocol version.
+
+### Using Compute Costs
+Compute costs different from gas costs are only be a temporary solution.
+Whenever we have introduce a compute cost, we as the community can discuss this publicly and find a solution to the specific problem together.
+
+In the best case, we find technical optimizations that allow us to remove compute cost without increasing gas costs.
+
+In other cases, the only solution is to increase the gas cost.
+But the dApp developers who are affected by this change should have a chance to voice their opinion, suggest alternatives, and implement necessary changes before the gas cost is increased.
+
+## Reference Implementation
+
+The reference implementation can be found at https://github.com/near/nearcore/pull/8426.
+
+The Compute Cost is a numeric value represented in time units.
+
+- [ ] Which type will we use? Which units will we use?
```

Comment:
Yeah we should decide and fill this in before marking this as ready for review.

We could use well-known types from the Rust ecosystem ([std](https://doc.rust-lang.org/std/index.html)::[time](https://doc.rust-lang.org/std/time/index.html)::[Duration](https://doc.rust-lang.org/std/time/struct.Duration.html#)) but probably just a `u64` (potentially with a strongly-typed wrapper) is better suited.

One reason not to use std Duration is it only has nanosecond granularity.
Current gas costs are in the granularity of millionth parts of a nanosecond, aka femtosecond (fs). I think the lowest value today is 98'562 fs. Therefore, ns granularity wouldn't be enough.

It feels like fs are a bit too granular and picoseconds (ps) (10^-12) would be granular enough. But to keep things in a simple relation to gas costs, I think we should just keep it as fs granularity (10^-15). A `u64` can then represent up to 18446 seconds ~ 5 hours. That seems reasonable to me.

Reply:
I agree with matching compute costs with gas cost scales and using u64. Updated the section to describe this.